Added Initial Quality Assurance guidelines/flow (#94)
langleyfoxall/technologies: quality-assurance/core.md

* Appropriate validation for different inputs (phone number, postcode)
* The forms are not too long &amp; and are grouped logically
* The user is alerted of the form lifecycle (success, error)
* When modals are used in the context of updates, ensure inputs are correctly pre populated.
* Autofill is disabled on inputs where appropriate
* Inputs have correct type property (emails have email type, phone numbers have tel type)

**Modals**
* Modals are animated on both display and dismiss
* Modals are used where appropriate over pages

**Formatting**
* Dates are formatted correctly for the appropriate audience.
* Monetary values are formatted correctly for the appropriate audience
* Phone numbers are formatted correctly for the appropriate audience

**Pagination**
* UI correctly indicates current page
* UI shows at least 2 page tabs ahead and below the current page

**Login &amp; Auth**
* Users can only access pages they are permitted access to.
* Authentication when they are logged in
* Login, registration (if applicable) and logout works as expected.
* Reset password works as expected

**Registration**
* Email addresses are verified &amp; unique (if applicable)

> * Passwords meet minimum standards (min 8 chars)

Reviewer: Should we enforce usage of https://github.com/langleyfoxall/laravel-nist-password-rules?

Reviewer 2: We could do something similar to the following.

```suggestion
* Passwords meet minimum standards, as defined by [NIST Special Publication 800-63B section 5](https://pages.nist.gov/800-63-3/sp800-63b.html#sec5). If using Laravel, ensure the [Laravel NIST Password Rules](https://github.com/langleyfoxall/laravel-nist-password-rules) is integrate.
```

Author: I'm against blanket enforcing the NIST rules due to their complexities and it would cause a significant barrier to entry for social apps (registration processes) and uptake to those platforms.

For that reason I think a lower level as a minimum (8 Chars) is enough, but its _preferable_ in our enterprise software to opt for NIST specification.

Reviewer 2: ```suggestion
* For enterprise software, passwords should preferably meet standards defined by [NIST Special Publication 800-63B section 5](https://pages.nist.gov/800-63-3/sp800-63b.html#sec5). If using Laravel, ensure the [Laravel NIST Password Rules](https://github.com/langleyfoxall/laravel-nist-password-rules) is integrated.
```

@wheatleyjj New suggestion based on your feedback.

> * Passwords meet minimum standards (min 8 chars)

Reviewer 2: ```suggestion
* For enterprise software, passwords should preferably meet standards defined by [NIST Special Publication 800-63B section 5](https://pages.nist.gov/800-63-3/sp800-63b.html#sec5). If using Laravel, ensure the [Laravel NIST Password Rules](https://github.com/langleyfoxall/laravel-nist-password-rules) is integrated.
```

@wheatleyjj New suggestion based on your feedback.

* Passwords should not have maximum length restrictions (allow at least 255 characters).

**Emails**
* Emails look good on small form factor devices
* Ensure correct subject &amp; sender name
* Correct email is sent, information is conveyed appropriately
* Emails are received in a timely manner
* Emails come from the appropriate email (often the clients)
* Links point to the correct URIs

**Push Notifications**
* Push notifications are received in a timely manner
* They show correct &amp; relevant information

**Permissions Roles**
* Ensure permissions &amp; roles match up to possible actions (only admins can create users etc)

### Whitebox
**General**
* Filenames must follow standard coding conventions (UserController.php)
* Classes must follow standard coding conventions (UserController)
* Variables should follow camelCase &amp; be descriptive (myName)

> * All long running operations (at your discretion) are queued

Reviewer: Should we also include that mail should be sent through a queue?

> * APIs follow REST architecture as closely as possible and implement our [API Response](https://github.com/langleyfoxall/helpers-laravel#apiresponse)

Reviewer: ```suggestion
* APIs follow REST architecture as closely as possible and implement our [API Response](https://github.com/langleyfoxall/helpers-laravel#apiresponse)
* API responses hide sensitive information outside of what is absolutely necessary.
```

Thoughts on this?

Reviewer 3: I was more referring to the Laravel helpers section, but you didn't write that so it was more a comment on the actual PR 😄

* Third party API integrations that can &amp; will be re-used across projects should be implemented as external packages, 
assuming no existing open-source integration exists, to allow for independent testing and code reuse if appropriate.
* Ensure third party API keys (e.g. SES keys) are not present within .env.example and other relevant files outside of the 
production / development environments.
* If Travis CI is being used, ensure third party API keys are input directly into Travis CI's website relevant to the 
project and make sure "Display value in build log" is not checked.
* Log viewers are hidden behind a protected route

**Controllers**
* Be on the lookout for duplicated code - recommend refactoring into helper classes or models if so
* Ensure all requests are validated either through a validator or custom request
* Errors are handled correctly in try catch blocks
* Responses are returned in a consistent &amp; logical format
* Only appropriate &amp; relevant information is returned
* When returning JSON ensure the returned JSON conforms to the [JSON spec](http://json-schema.org/specification.html)

> * Appropriate status codes are set

Reviewer 4: ```suggestion
* Appropriate status codes are set
  * `201` - For creation API responses
  * `202` - For sending requests that are successful to a third party API
  * `401` - For unauthenticated users trying to access protected areas
  * `403` - For normal users trying to access admin areas
  * `422` - For missing required data in the request body
  * Continue reading [here](https://httpstatuses.com/)
```

Might it be worth giving examples or linking off to a resource with descriptions on HTTP status codes? I personally like [httpstatuses.com](https://httpstatuses.com/).

Discussing what status codes to be used where/when is probably something that can defined in more detail in another document.

Author: Interesting that you've missed `204` and `200` off there as common ones.

Unsure whether this is the right document for them - thumbs 👍 👎

Reviewer 3: I think having examples of the right status codes is the right way, enforcing them I think puts a blanket requirement on something that isn't always black and white.

What is we're talking to a third party and then creating a resource, would we go with `201`, `202` or just plain old `200`?

I think it needs it's own discussion about the correct status codes to use in different circumstances.

Reviewer 4: > Unsure whether this is the right document for them - thumbs 👍 👎

> I think it needs it's own discussion about the correct status codes to use in different circumstances.

As mentioned, I think that status codes could be it's own discussion and document but we should at least point towards a good resource for status codes, as I'm sure some people wouldn't bother googling and just put `200`s for everything.

Reviewer 2: I think example should be pulled into their own markdown file, linked to from here.

Author: Executive decision to leave this for another section of the tech repo. This doesn't quite relate to Quality Assurance enough for me.

* Large tables that are queried often are indexed in order to keep database speed up

## Processes
### Automated Testing

Automated testing should be implemented in a way to confirm functionality specified by the client.
To this end, automated tests should be created to confirm the project user stories and related 
business logic is satisfied and correct. 

#### Unit Testing

Our current policy is that unit tests should be used for:

* Testing of project business logic / rules, that are able to be sufficient abstracted
* Testing of packages, including our open source work

#### Integration, Browser &amp; End-to-End Testing

Integration, browser and end-to-end tests should be used for:

* Testing of project functionality related to client specified user stories
* Testing of critical path functionality, such as the register, login, and checkout process of an
ecommerce site

Since end-to-end tests are slower than browser tests and browser tests are slower than integration 
tests, you should ideally aim to use the fastest possible test type, if appropriate.
### Manual Testing
#### 1 - Developer Testing

> During the development of the project, developers will create `feature/*` branches for all features going into the software.

Reviewer: I don't think this statement is needed since we already have this highlighted in our git flow documentation.

Reviewer 2: @ash123456789 Agreed. 👍 

```suggestion
During the development of the project, developers will create branches as described in our git work flow documentation.
```

After the initial feature development - the developer should manually test their feature after writing passing 
[automation tests](#automated-testing). This phase should also include a check that ensures the feature does not infringe
any of the items listed in the ["What makes a good web application?"](#what-makes-a-good-web-application) section. 

> Once the feature has the thumbs up from the developer - that is when the pull request into the `master` branch should be

Reviewer: Again, this is also highlighted in our git flow documentation and I don't think this is needed here.

# 🚦 Langley Foxall Quality Assurance

When building applications that run an entire business end-to-end - it's important that we adhere to high quality
assurance standards in order to consistently deliver a high quality software project.

There are different parts/processes to quality assurance as a whole, and these are not limited to just manual testing and
automated testing.

*   [What makes a good web application?](#what-makes-a-good-web-application)
    *   [Blackbox](#blackbox)  
    *   [Whitebox](#whitebox)  
*   [Processes](#processes)
    *   [Automated Testing](#automated-testing)
    *   [Manual Testing](#manual-testing)

## What makes a good web application?
The following is a checklist of things to think about both during development of a product and during the quality 

> assurance phase.

Reviewer 2: ```suggestion
assurance phase. They should be considered guidelines, and not a strict set of requirements, as not all points will be applicable or appropriate for all applications.
```

* The user is alerted of the form lifecycle (success, error)
* When modals are used in the context of updates, ensure inputs are correctly pre populated.
* Autofill is disabled on inputs where appropriate
* Inputs have correct type property (emails have email type, phone numbers have tel type)

**Modals**
* Modals are animated on both display and dismiss
* Modals are used where appropriate over pages

**Formatting**
* Dates are formatted correctly for the appropriate audience.
* Monetary values are formatted correctly for the appropriate audience
* Phone numbers are formatted correctly for the appropriate audience

**Pagination**
* UI correctly indicates current page
* UI shows at least 2 page tabs ahead and below the current page

**Login &amp; Auth**
* Users can only access pages they are permitted access to.
* Authentication when they are logged in
* Login, registration (if applicable) and logout works as expected.
* Reset password works as expected

**Registration**
* Email addresses are verified &amp; unique (if applicable)
* Passwords meet minimum standards (min 8 chars)
* Passwords should not have maximum length restrictions (allow at least 255 characters).

**Emails**

> * Emails look good on small form factor devices

Reviewer 5: Might be worth mentioning ensuring the HTML and CSS in the email works on the popular clients if custom HTML and CSS is being written. There are good resources such as [this guide](https://www.campaignmonitor.com/css/) which lets you search by email clients.

* Testing of project functionality related to client specified user stories
* Testing of critical path functionality, such as the register, login, and checkout process of an
ecommerce site

Since end-to-end tests are slower than browser tests and browser tests are slower than integration 
tests, you should ideally aim to use the fastest possible test type, if appropriate.
### Manual Testing
#### 1 - Developer Testing
During the development of the project, developers will create `feature/*` branches for all features going into the software.
After the initial feature development - the developer should manually test their feature after writing passing 
[automation tests](#automated-testing). This phase should also include a check that ensures the feature does not infringe
any of the items listed in the ["What makes a good web application?"](#what-makes-a-good-web-application) section. 

Once the feature has the thumbs up from the developer - that is when the pull request into the `master` branch should be
made. 

The developers should log any bugs found in this phase into JIRA to improve their ratio of bugs they find VS bugs that
get through to the QA team/customer. 

> #### 2 - Sprint Testing (On the master branch)

Reviewer 2: A possible problem  count occur here if QA is testing on the `master` branch, and development continues on with development of the next sprint in the mean time.

Reviewer 3: @wheatleyjj What about testing on staging? We should be aiming to push a version to staging at the end of any sprint so it makes sense to do QA on that branch.

Reviewer: I agree with @AlexCatch about testing on staging.

Reviewer 2: I've worked on project previously with this kind of QA process in place. 

At the start of a sprint, a `sprint-1` branch would be created. After the sprint is over, QA would begin testing on `sprint-1`. At the same time a `sprint-2` branch would be created (from `sprint-1`) to allow developers to continue development immediately.

Once QA was happy with `sprint-1` it would be merged into `mater`, and the process would repeat.

Author: @ash123456789 @AlexCatch I think that testing on the customers environment would cause annoyances to the client if they're using it regularly. 

Ways around this are sprint branches as I and @DivineOmega suggested or having a development environment - but that comes with another layer of grief. I'm also thinking experienced QA testers will have good knowledge of automation tests so would be able to look into both the Blackbox/Whitebox sections of this document.

Thumbs up if you agree?

Reviewer 2: @DivineOmega Agree with the sprint branches and/or development server idea. :+1:

assurance standards in order to consistently deliver a high quality software project.

There are different parts/processes to quality assurance as a whole, and these are not limited to just manual testing and
automated testing.

*   [What makes a good web application?](#what-makes-a-good-web-application)
    *   [Blackbox](#blackbox)  
    *   [Whitebox](#whitebox)  
*   [Processes](#processes)
    *   [Automated Testing](#automated-testing)
    *   [Manual Testing](#manual-testing)

## What makes a good web application?
The following is a checklist of things to think about both during development of a product and during the quality 
assurance phase.

### Blackbox
**General**
* All UI scales well and looks good on small form factor devices
* All errors are handled and UI reacts appropriately
* Errors are clear in meaning &amp; reflect what happened
* Activity indicators are used when loading data asynchronously
* Empty information is handled correctly in UI (optional fields etc)
* Long running operations give a indication of progress if possible
* Breadcrumbs are used where appropriate

**Navbar**
* UI reflects the current active page

> * Icon is appropriate for the page

Reviewer 4: ```suggestion
* An icon that is appropriate for the page (if applicable)
```

**Forms**
* Invalid inputs correctly trigger validation errors
* All form inputs have associated labels
* Appropriate validation for different inputs (phone number, postcode)
* The forms are not too long &amp; and are grouped logically
* The user is alerted of the form lifecycle (success, error)
* When modals are used in the context of updates, ensure inputs are correctly pre populated.
* Autofill is disabled on inputs where appropriate

> * Inputs have correct type property ( emails have email type, phone numbers have tel type)

Reviewer 4: ```suggestion
* Inputs have correct type property (emails have email type, phone numbers have tel type)
```

**Modals**
* Modals are animated on both display and dismiss
* Modals are used where appropriate over pages

**Formatting**
* Dates are formatted correctly for the appropriate audience.
* Monetary values are formatted correctly for the appropriate audience
* Phone numbers are formatted correctly for the appropriate audience

**Pagination**
* UI correctly indicates current page
* UI shows at least 2 page tabs ahead and below the current page

**Login &amp; Auth**

> * User can only access pages that require 

Reviewer: ```suggestion
* Users can only access pages they are permitted access to.
```

* Authentication when they are logged in

> * Login &amp; Logout works as expected

Reviewer: ```suggestion
* Login, registration (if applicable) and logout works as expected.
```

* Reset password works as expected

**Registration**
* Email addresses are verified &amp; unique (if applicable)
* Passwords meet minimum standards (min 8 chars)
* Passwords should not have maximum length restrictions (allow at least 255 characters).

**Emails**
* Emails look good on small form factor devices
* Ensure correct subject &amp; sender name
* Correct email is sent, information is conveyed appropriately
* Emails are received in a timely manner
* Emails come from the appropriate email (often the clients)

> * Links point to the correct address

Reviewer 4: ```suggestion
* Links point to the correct URIs
```

**Push Notifications**
* Push notifications are received in a timely manner
* They show correct &amp; relevant information

**Permissions Roles**
* Ensure permissions &amp; roles match up to possible actions (only admins can create users etc)

### Whitebox
**General**

> * Filenames must follow standard coding conventions (UsersController.php)

Reviewer 4: ```suggestion
* Filenames must follow standard coding conventions (UserController.php)
```

I prefer to have singular controller names, to continue the naming convention of having singular models

Reviewer 3: It's all opinion based, I prefer the plural but it looks like on the Laravel docs they use singular naming so that makes sense.

> * Classes must follow standard coding conventions (UsersController)

Reviewer 4: ```suggestion
* Classes must follow standard coding conventions (UserController)
```

* Variables should follow camelCase &amp; be descriptive (myName)
* All long running operations (at your discretion) are queued

> * APIs follow REST architecture as closely as possible

Reviewer 4: ```suggestion
* APIs follow REST architecture as closely as possible and implement our [API Response](https://github.com/langleyfoxall/helpers-laravel#apiresponse)
```

* Auth middleware is setup &amp; working correctly
* Event system is used for method side effects (Audit trail etc)
* Laravel Notifications are used to notify the user of system events (subscription expired, new friend request etc)
* Sensitive values are either hashed or encrypted (when required by law or common sense)
* Third party API integrations that can &amp; will be re-used across projects should be implemented as external packages, 
assuming no existing open-source integration exists, to allow for independent testing and code reuse if appropriate.
* Ensure third party API keys (e.g. SES keys) are not present within .env.example and other relevant files outside of the 
production / development environments.
* If Travis CI is being used, ensure third party API keys are input directly into Travis CI's website relevant to the 
project and make sure "Display value in build log" is not checked.
* Log viewers are hidden behind a protected route

**Controllers**

> * Ensure each controller function does not go over 250 lines

Reviewer 3: This should be the entire class and not just function, a single function should really never go over 50 lines.

Author: > This should be the entire class and not just function, a single function should really never go over 50 lines.

This is your document 🤕

Reviewer 3: Just pointing out a typo.

* Be on the lookout for duplicated code - recommend refactoring into helper classes or models if so
* Ensure all requests are validated either through a validator or custom request
* Errors are handled correctly in try catch blocks
* Responses are returned in a consistent &amp; logical format
* Only appropriate &amp; relevant information is returned

> * When returning JSON ensure the returned JSON conforms to the JSON spec (http://json-schema.org/specification.html)

Reviewer 4: ```suggestion
* When returning JSON ensure the returned JSON conforms to the [JSON spec](http://json-schema.org/specification.html)
```

* Appropriate status codes are set

**Emails**
* Emails use the SES driver unless agreed beforehand

> * All emails are queued up for sending

Reviewer 4: This could be seen as clashing with an earlier point on queues:
> * All long running operations (at your discretion) are queued

Reviewer 3: I guess they are essentially making the same point, it could be up to interpretation whether an email constitutes `a long running operation` though so it's better to be explicit I feel.

* SQL data types are appropriate
* Monetary values should either be stored in pence as an INTEGER or as a DECIMAL - Never a FLOAT
* Large tables that are queried often are indexed in order to keep database speed up

## Processes
### Automated Testing

Automated testing should be implemented in a way to confirm functionality specified by the client.
To this end, automated tests should be created to confirm the project user stories and related 
business logic is satisfied and correct. 

#### Unit Testing

Our current policy is that unit tests should be used for:

* Testing of project business logic / rules, that are able to be sufficient abstracted
* Testing of packages, including our open source work

#### Integration, Browser &amp; End-to-End Testing

Integration, browser and end-to-end tests should be used for:

* Testing of project functionality related to client specified user stories
* Testing of critical path functionality, such as the register, login, and checkout process of an
ecommerce site

Since end-to-end tests are slower than browser tests and browser tests are slower than integration 
tests, you should ideally aim to use the fastest possible test type, if appropriate.

> e

Reviewer 4: ```suggestion

```

Reviewer 4: Is this in reference to [eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee](https://github.com/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee)?
